### ISSUE TITLE
Receiving an ipc channel uses an fd

### ISSUE THREAD

**Issue by asajeffrey** [servo/ipc-channel#240]
When ipc channels are sent over ipc oin unix, their fds are sent using `sendmsg`, and the matching `recvmsg` creates a new fd. This means that even programs which use a fixed number of channels can end up using an unbounded number of fds. For example:
```rust
fn main() {
   let (send1, recv1) = ipc::channel::<IpcSender<bool>>().unwrap();
   let (send2, recv2) = ipc::channel::<bool>().unwrap();
   thread::spawn(move || {
       let mut senders = vec![];
       while let Ok(send2) = recv1.recv() {
           let _ = send2.send(true);
           // The fd is private, but this transmute lets us get at it
           // let fd: &std::sync::Arc<u32> = unsafe { std::mem::transmute(&send2) };
	   // println!("fd = {}", *fd);
	   // Stop the ipc channel from being dropped
	   senders.push(send2);
       }
   });
   for _ in 0..10000 {
       let _ = send1.send(send2.clone());
       let _ = recv2.recv();
   }
}
```
runs out of fds even though it only uses two ipc channels.

**Comment by gterzian** [comment]
I wonder how relevant the above example is in practice.

If a program distributes many clones of senders from one process to many different processes, it will be because the program intends to setup a communication channel between that one process and all others. 

In such a scenario, does the number of processes the machine would be able to manage not provide an even stronger limitation on the feasibility of the design, versus the number of fds? 

And would there be a way to setup communication with more than one process while avoiding using an fd for each?

If the program only needs to  setup a communication channel between two processes to handle many operations, and that is done by distributing the clone of a sender for each operation, could the communication be refactored to use one sender only, thereby avoiding sending clones for each operation?

**Comment by asajeffrey** [comment]
We could ask people to refactor their programs, but that's not very satisfactory, as it means the cost model for ipc is very different than it is for mpsc, where sending a channel is cheap.

**Comment by gterzian** [comment]
> the cost model for ipc is very different than it is for mpsc, where sending a channel is cheap.

That's simply a fact. The question is what is the right solution?

I can think of two options:

1. The maintainers of ipc-channel can commit to giving developers the illusion that using an ipc-channel comes with the same cost/benefit model as that of a multithreaded channel. Could this make the implementation more complicated then it needs to be?

2. ipc-channel can have an API that is similar to other channels, yet end-users have to be aware that the underlying cost/benefit model is not the same. Could this be a way to keep the library lean and mean, leaving end-users to adopt various additional abstraction on top of it to fit their use-case?

As it currently stands, ipc-channel is good-enough for most use-cases. And when a project runs into specific problems, it can address that with a refactoring. Could trying to fit the library to every use-case instead be a recipe for an over-complicated implementation?

I think it's Python that has this policy of not allowing any optimization for a specific case that could slow down the base case(or make its implementation more complicated?).

**Comment by gterzian** [comment]
I think we also need to acknowledge that in practice, a sender is ipc'ed almost always to implement a blocking ipc call, which requires creating a channel so as to have a dedicated receiver to block on the reply. 

The sender and receiver could be stored somewhere, instead of creating a channel for each call, and a clone of the sender could be ipc'ed. In that case, we can imagine a situation where no new fd's are created if the fd is somehow cached by the process receiving the sender clone. 

However that doesn't work in all cases, in fact I don't think it would work in most cases where blocking ipc call are implemented. The reason is a situation where many distinct object need to make many distinct blocking ipc call. In such a situation, each object requires it's own channel, because those objects can exist in parallel of each other. 

Think a websocket or an XHR request in Servo. They either each need a new channel, or they would require some routing layer using a single channel(per script-process). In such a setup, which I think is the common case, channels cannot be recyled without a routing glue layer, which means caching fds wouldn't actually solve the problem in practice, however a project specific routing layer that would simply use an ipc-channel(that doesn't cache fds), would solve the problem.

**Comment by glyn** [comment]
I don't think this issue can be fixed without a redesign of the socket-based Unix platform support or a change to the external API (to rule out the scenarios that cause the problem).

For example, a client could obtain senders for arbitrarily many distinct server processes and send them to a single server which would then need a distinct file descriptor to communicate with each server process.

That said, Servo seems to have tolerated the issue for over five years, so maybe it should be declared "working as designed"?

**Comment by mrobinson** [comment]
This is fairly serious issue we are struggling with in Servo. See https://github.com/servo/servo/issues/23905.

**Comment by glyn** [comment]
Please see https://github.com/servo/ipc-channel/issues/396 for a prototype of a proposed solution.
